test/modeling_library/distributions.jl

```diff
@@ -233,6 +233,34 @@ end
     @test isapprox(actual[3][2, 2], finite_diff_mat_sym(f, args, 3, 2, 2, dx))
 end
 
+@testset "dirichlet" begin
+    x = dirichlet([1., 1., 1., 1.])
+    @test length(x) == 4
+    @test isapprox(sum(x), 1.)
+
+    # bounds checking
+    @test logpdf(dirichlet, [0., 0], [1., 1.]) == -Inf
+    @test logpdf(dirichlet, [1., 1.], [1., 1.]) == -Inf
+    @test logpdf(dirichlet, [2., -1], [1., 1.]) == -Inf
+    @test logpdf(dirichlet, [.5, .5], [-1., 1.]) == -Inf
+    @test logpdf(dirichlet, [.5, .5], [-1., 1.]) == -Inf
+    @test logpdf(dirichlet, [0., 1], [1., 1.]) != -Inf
+
+    @test isapprox(logpdf(dirichlet, [.01, .99], [2., 2.]),
+                   Distributions.logpdf(Distributions.Dirichlet([2., 2.]), [.01, .99]))
+    @test isapprox(logpdf(dirichlet, [.01, .99], [1., 4.]),
+                   Distributions.logpdf(Distributions.Dirichlet([1., 4.]), [.01, .99]))
+    @test isapprox(logpdf(dirichlet, [.01, .99], [.01, .01]),
+                   Distributions.logpdf(Distributions.Dirichlet([.01, .01]), [.01, .99]))
+
+    f = (x, alpha) -> logpdf(dirichlet, x, alpha)
+    args = ([.5, .5], [1., 2.])
+    actual = logpdf_grad(dirichlet, args...)
+
+    @test isapprox(actual[2][1], finite_diff_vec(f, args, 2, 1, dx))
+    @test isapprox(actual[2][2], finite_diff_vec(f, args, 2, 2, dx))
```

Comment:
For finite differencing for `N >= 3`, maybe we can write a version of `finite_diff_vec` that maintains the constraint of remaining on the simplex? E.g. if you increase the probability vector at position `i` by `dx`, then decrease the probability vector at positions `j != i` by `dx / N-1`.

Also, am I right to say that these test cases check whether perturbations to `alpha` are resulting in the right gradients? If so, I would either rewrite the code or add comments to make that a little clearer.

Reply:
So that was what I tried at first, and the problem with that approach is that you're now modifying every entry of the vector x, meaning that the finite difference you get out won't approximate the partial derivative - which is by definition, what happens when you hold every element besides `i` constant.

Reply:
Hmm, don't they end up being equal though, for sufficiently small differences? Or is that not the case?

Reply:
Another option - what about if we instead test the approximate correctness of the gradient of `logpdf(f(x))`, where `f` is a function that maps from R^n to the simplex, such as `softmax`? Since the gradient of `softmax` is known, we can multiply it by the output of `logpdf_grad` to get the full gradient via the chain rule. And then we can check that value against finite differencing of `logpdf(softmax(x))`.